src/traces/bar/cross_trace_calc.js

```diff
@@ -56,7 +56,7 @@ function crossTraceCalc(gd, plotinfo) {
 function setGroupPositions(gd, pa, sa, calcTraces) {
     if(!calcTraces.length) return;
 
-    var barmode = gd._fullLayout.barmode || gd._fullLayout.waterfallmode;
+    var barmode = (gd._fullLayout.barmode !== undefined) ? gd._fullLayout.barmode : gd._fullLayout.waterfallmode;
```

Comment:
Hmm. This seems wrong. What happens on subplots with both bar and waterfall traces? In the case, both `barmode` and `waterfallmode` are present in `fullLayout` and here the waterfall traces would be computed using `barmode` instead of `waterfallmode`.

Please have a look at how `barpolar` handles this problem:

https://github.com/plotly/plotly.js/blob/7b2946a4c2c6cd6ab5691e5a1fc37b8809831287/src/traces/barpolar/calc.js#L76-L104

where we don't call `Bar.crossTraceCalc` direction, but instead we call `setGroupPositions` with a mock `gd` object containing the correct `*mode`, `*groupgap` and `*gap` attribute.